{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working directory: /mnt/disks/store/ist_benchmarking\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "wd = os.getcwd()\n",
    "print (f'working directory: {wd}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import glob\n",
    "import matplotlib.colors as colors\n",
    "\n",
    "from st_utils import df_2_gdf\n",
    "from st_utils import get_gdf_core\n",
    "from st_utils import get_gene_type\n",
    "from st_utils import transcript_loader\n",
    "from st_utils import correct_tissue_names, correct_platform_panel\n",
    "from st_utils import name_parser\n",
    "\n",
    "\n",
    "from constants import SCALING_FACTOR_DICT\n",
    "from constants import CORE_RADIUS_DICT\n",
    "from constants import POINTS_SRC_DICT\n",
    "from constants import UNIQUE_GENES_DICT\n",
    "from constants import SAMPLES\n",
    "\n",
    "pd.set_option('display.max_rows',999)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)\n",
    "plt.rcParams['figure.dpi']=250\n",
    "\n",
    "import warnings\n",
    "if not sys.warnoptions:\n",
    "    warnings.simplefilter(\"ignore\")\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "\n",
    "with open(f'{wd}/data/common_cores.pkl', 'rb') as f:\n",
    "    matching_cores = pickle.load(f)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def xenium_fdr(sample, neg_metric, common_cores=matching_cores):\n",
    "\n",
    "    modality, panel, tma = name_parser(sample)\n",
    "    # Core information: including tissue_type, tumor ...\n",
    "    csv_sample_info = f'{wd}/data/Sample_Info_{tma.upper()}_TMA - Sheet1.csv'\n",
    "\n",
    "    df_meta = pd.read_csv(csv_sample_info)\n",
    "    df_meta = correct_tissue_names(sample, df_meta)\n",
    "    df_meta['core'] = df_meta['core'].astype('int')\n",
    "    df_meta['core'] += (300 if 'tumor2' in sample else 0) + (200 if 'normal' in sample else 0)\n",
    "    df_meta = df_meta.loc[df_meta['core'].isin(matching_cores)]\n",
    "    df_meta['core'] = df_meta['core'].astype('str')\n",
    "\n",
    "    # Get cell-level data\n",
    "    df_cell = pd.read_parquet(f'{wd}/data/cell_level_csv/{sample}_cell_level.parquet.gzip', engine='pyarrow')\n",
    "    df_cell = correct_tissue_names(sample, df_cell)\n",
    "\n",
    "\n",
    "    df_cell['core'] = df_cell['core'].astype('int')\n",
    "    df_cell['core'] += (300 if 'tumor2' in sample else 0) + (200 if 'normal' in sample else 0)\n",
    "    df_cell = df_cell.loc[df_cell['core'].isin(matching_cores)]\n",
    "    df_cell['core'] = df_cell['core'].astype('str')\n",
    "\n",
    "    # customize aggregation for each col\n",
    "    df_metric = df_cell.groupby(by = ['core'], as_index=False).agg({\n",
    "        'cell_id':'count',\n",
    "        'transcript_counts':'sum',\n",
    "        'control_probe_counts':'sum',\n",
    "        'control_codeword_counts':'sum',\n",
    "        'unassigned_codeword_counts':'sum',\n",
    "        'total_counts':'sum',\n",
    "        })\n",
    "\n",
    "    df_metric = df_metric.rename(columns={'cell_id':'cell_count'})\n",
    "\n",
    "    # calculate transcripts_per_cell\n",
    "    df_metric['transcripts_per_cell'] = df_metric.apply(\n",
    "    lambda row: round(row['total_counts']/row['cell_count'],1),\n",
    "    axis=1)\n",
    "\n",
    "    # get tissue_type\n",
    "    df_metric = pd.merge(df_metric, df_meta[['core','tissue_type']], on='core', how='inner')\n",
    "\n",
    "    # Calculate FDR\n",
    "    df_metric['fdr'] = df_metric.apply(\n",
    "        lambda row: (np.true_divide(row['control_probe_counts'],row['total_counts'])\n",
    "                    ) * (UNIQUE_GENES_DICT[sample]['gene'] / UNIQUE_GENES_DICT[sample][neg_metric]) * 100, axis=1)\n",
    "\n",
    "    df_metric = df_metric.sort_values(by=['core'])\n",
    "\n",
    "    return df_metric\n",
    "\n",
    "\n",
    "def merscope_fdr(sample, neg_metric,common_cores=matching_cores):\n",
    "\n",
    "    modality, panel, tma = name_parser(sample)\n",
    "    # Core information: including tissue_type, tumor ...\n",
    "    csv_sample_info = f'data/Sample_Info_{tma.upper()}_TMA - Sheet1.csv'\n",
    "\n",
    "    df_meta = pd.read_csv(csv_sample_info)\n",
    "    df_meta = correct_tissue_names(sample, df_meta)\n",
    "    df_meta['core'] = df_meta['core'].astype('int')\n",
    "    df_meta.core = df_meta.core + 200 if 'normal' in sample else df_meta.core\n",
    "    # df_meta = df_meta.loc[df_meta['core'].isin(matching_cores)]\n",
    "    df_meta['core'] = df_meta['core'].astype('str')\n",
    "\n",
    "    # Get cell-level data\n",
    "    df_cell = pd.read_parquet(f'data/cell_level_csv/{sample}_cell_level.parquet.gzip', engine='pyarrow')\n",
    "    df_cell = correct_tissue_names(sample, df_cell)\n",
    "\n",
    "    df_cell['core'] = df_cell['core'].astype('int')\n",
    "    df_cell.core = df_cell.core + 200 if 'normal' in sample else df_cell.core\n",
    "    # df_cell = df_cell.loc[df_cell['core'].isin(matching_cores)]\n",
    "    df_cell['core'] = df_cell['core'].astype('str')\n",
    "\n",
    "    # Get gdf_core\n",
    "    # core centroid data x/y in pixel\n",
    "    csv_points = glob.glob(f'data/{sample}/*points.csv')[0]\n",
    "\n",
    "    # 1 pixel = scaling_factor um\n",
    "    scaling_factor = SCALING_FACTOR_DICT[sample]\n",
    "    radius_um = CORE_RADIUS_DICT[sample]\n",
    "\n",
    "    # Get the gdf of core information with geometry (buffer from the core centroid)\n",
    "    gdf_core= get_gdf_core(csv_points, csv_sample_info, scaling_factor, radius_um, points_src=POINTS_SRC_DICT[sample])\n",
    "    gdf_core = correct_tissue_names(sample, gdf_core)\n",
    "    gdf_core['core'] = gdf_core['core'].apply(lambda x: str(x).zfill(3))\n",
    "\n",
    "    # Get transcript data\n",
    "    df_t = transcript_loader(f'data/{sample}')\n",
    "    df_t['calc_id']  = df_t['cell_id'].apply(lambda x: x.split('_region')[0]).astype(float)\n",
    "    df_t = df_t.loc[df_t['calc_id']>0]\n",
    "\n",
    "    gdf_t = df_2_gdf(df_t, 'global_x', 'global_y', crs=\"EPSG:4326\", drop_xy=False)\n",
    "\n",
    "    extra_variable = []\n",
    "\n",
    "    gdf_join = gdf_core[\n",
    "        ['core','tissue_type','geometry']\n",
    "        ].sjoin(gdf_t, how='right', op='intersects'\n",
    "                ).drop(columns='index_left').dropna(subset=['core'])\n",
    "    \n",
    "    gdf_join['core'] = gdf_join['core'].apply(lambda x: str(x).zfill(3))\n",
    "    gdf_join['gene_type'] = gdf_join['gene'].apply(lambda x: get_gene_type(x))\n",
    "\n",
    "    tts = []\n",
    "    merscope_fdrs = []\n",
    "\n",
    "    for tt in gdf_join.tissue_type.unique():\n",
    "        df_tt = gdf_join.loc[gdf_join['tissue_type']==tt]\n",
    "        gene_count = {**Counter(df_tt.gene_type)}\n",
    "        print (tt, gene_count)\n",
    "        merscope_fdr = (np.true_divide(gene_count.get(neg_metric,0), gene_count.get(neg_metric,0)+ gene_count.get('gene',0)) * (UNIQUE_GENES_DICT[sample]['gene'] / UNIQUE_GENES_DICT[sample][neg_metric]) * 100 )\n",
    "        tts.append(tt)\n",
    "        merscope_fdrs.append(merscope_fdr)\n",
    "\n",
    "\n",
    "    print (f'df_t len: {len(df_t)}, df_join len:{len(gdf_join)}')\n",
    "\n",
    "    df_metric = pd.DataFrame({'tissue_type':tts, 'fdr':merscope_fdrs})\n",
    "\n",
    "    return df_metric\n",
    "\n",
    "\n",
    "def cosmx_fdr(sample, neg_metric,common_cores=matching_cores):\n",
    "\n",
    "    # Get cell-level data\n",
    "    df_cell = pd.read_parquet(f'{wd}/data/cell_level_csv/{sample}_cell_level.parquet.gzip', engine='pyarrow')\n",
    "    df_cell = correct_tissue_names(sample, df_cell)\n",
    "    df_cell['core'] = df_cell['core'].astype('int')\n",
    "    df_t = transcript_loader(f'{wd}/data/{sample}')\n",
    "    \n",
    "    df =pd.merge(df_t, df_cell[['core', 'tissue_type', 'cell_id']], on='cell_id', how='left')\n",
    "\n",
    "    df['core'] += (300 if 'tumor2' in sample else 0) + (200 if 'normal' in sample else 0)\n",
    "    df = df.loc[df['core'].isin(matching_cores)]\n",
    "    df = df.dropna(subset=['core'])\n",
    "    print (f'ct/spot/join:{len(df_cell)}/{len(df_t)}/{len(df)}')\n",
    "\n",
    "    df['core'] = df['core'].apply(lambda x: str(x).zfill(3))\n",
    "    df['gene_type'] = df['gene'].apply(lambda x: get_gene_type(x))\n",
    "\n",
    "    tts = []\n",
    "    cosmx_fdrs = []\n",
    "\n",
    "    for tt in df.tissue_type.unique():\n",
    "        df_tt = df.loc[df['tissue_type']==tt]\n",
    "        gene_count = {**Counter(df_tt.gene_type)}\n",
    "        print (tt, gene_count)\n",
    "        cosmx_fdr = (np.true_divide(gene_count.get(neg_metric,0), gene_count.get(neg_metric,0)+ gene_count.get('gene',0)) * (UNIQUE_GENES_DICT[sample]['gene'] / UNIQUE_GENES_DICT[sample][neg_metric]) * 100 )\n",
    "        tts.append(tt)\n",
    "        cosmx_fdrs.append(cosmx_fdr)\n",
    "\n",
    "    print (f'df_t len: {len(df_t)}, df_join len:{len(df)}')\n",
    "\n",
    "    df_metric = pd.DataFrame({'tissue_type':tts, 'fdr':cosmx_fdrs})\n",
    "\n",
    "    return df_metric\n",
    "\n",
    "\n",
    "\n",
    "cmap = colors.LinearSegmentedColormap.from_list('nameofcolormap',['#0072B2','#D55E00'],gamma=2.0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reference: https://pages.10xgenomics.com/tch-2023-04-tech-lit-ra_g-p_xenium-performance-data-lp.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FDR Figure 1: Xenium (blank) vs. MERSCOPE(blank) vs. CosMX (sys control)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample: xenium_breast_htma\n",
      "sample: xenium_breast_normal\n",
      "sample: xenium_panhuman_htma\n",
      "sample: xenium_panhuman_normal\n",
      "sample: xenium_lung_htma\n",
      "sample: xenium_lung_normal\n",
      "sample: merscope_breast_htma\n",
      "data/merscope_breast_htma/region_0/\n",
      "data/merscope_breast_htma/region_1/\n",
      "data/merscope_breast_htma/region_2/\n",
      "data/merscope_breast_htma/region_3/\n",
      "data/merscope_breast_htma/region_4/\n",
      "data/merscope_breast_htma/region_5/\n",
      "data/merscope_breast_htma/region_6/\n",
      "data/merscope_breast_htma/region_7/\n",
      "data/merscope_breast_htma/region_8/\n",
      "17743227\n",
      "HNSCC {'gene': 1766270, 'blank': 12233}\n",
      "Mel {'gene': 1040225, 'blank': 5311}\n",
      "BrC {'gene': 2151252, 'blank': 10567}\n",
      "NSCLC {'gene': 2310812, 'blank': 15687}\n",
      "OvC {'gene': 2258259, 'blank': 6667}\n",
      "CRC {'gene': 1932167, 'blank': 11549}\n",
      "BlC {'gene': 1696510, 'blank': 10316}\n",
      "df_t len: 13823088, df_join len:13227825\n",
      "sample: merscope_breast_normal\n",
      "data/merscope_breast_normal/region_0/\n",
      "data/merscope_breast_normal/region_1/\n",
      "data/merscope_breast_normal/region_2/\n",
      "data/merscope_breast_normal/region_3/\n",
      "data/merscope_breast_normal/region_4/\n",
      "data/merscope_breast_normal/region_5/\n",
      "data/merscope_breast_normal/region_6/\n",
      "data/merscope_breast_normal/region_7/\n",
      "data/merscope_breast_normal/region_8/\n",
      "375204\n",
      "Marker {'gene': 24757, 'blank': 171}\n",
      "Skin {'gene': 690, 'blank': 8}\n",
      "Thyroid {'gene': 21808, 'blank': 48}\n",
      "Breast {'gene': 2266, 'blank': 25}\n",
      "Lymph node {'gene': 479, 'blank': 1}\n",
      "Spleen {'gene': 2384, 'blank': 106}\n",
      "Tonsil {'gene': 19247, 'blank': 73}\n",
      "Liver {'gene': 8339, 'blank': 40}\n",
      "Heart {'gene': 599, 'blank': 36}\n",
      "Kidney {'gene': 6319, 'blank': 10}\n",
      "Lung {'gene': 87, 'blank': 1}\n",
      "Colon {'gene': 1765, 'blank': 22}\n",
      "Ovary {'gene': 13033, 'blank': 18}\n",
      "Bladder {'gene': 1093, 'blank': 2}\n",
      "df_t len: 240203, df_join len:103427\n",
      "sample: merscope_lung_htma\n",
      "data/merscope_lung_htma/region_0/\n",
      "data/merscope_lung_htma/region_1/\n",
      "data/merscope_lung_htma/region_2/\n",
      "data/merscope_lung_htma/region_3/\n",
      "data/merscope_lung_htma/region_4/\n",
      "3251107\n",
      "HNSCC {'gene': 163404, 'blank': 2791}\n",
      "CRC {'gene': 390858, 'blank': 5038}\n",
      "Mel {'gene': 149335, 'blank': 1802}\n",
      "BrC {'gene': 243130, 'blank': 7917}\n",
      "NSCLC {'gene': 460749, 'blank': 5523}\n",
      "Tonsil {'gene': 43254, 'blank': 1877}\n",
      "Lymph node {'gene': 50300, 'blank': 670}\n",
      "OvC {'gene': 870973, 'blank': 7701}\n",
      "df_t len: 2480584, df_join len:2405322\n",
      "sample: merscope_lung_normal\n",
      "data/merscope_lung_normal/region_0/\n",
      "156656\n",
      "Spleen {'gene': 2603, 'blank': 244}\n",
      "Pancreas {'gene': 2775, 'blank': 43}\n",
      "Renal pelvis {'gene': 2469, 'blank': 36}\n",
      "Kidney {'gene': 5235, 'blank': 18}\n",
      "Tonsil {'gene': 6960, 'blank': 89}\n",
      "Thyroid {'gene': 9186, 'blank': 53}\n",
      "Skin {'gene': 443, 'blank': 3}\n",
      "Lung {'gene': 76, 'blank': 1}\n",
      "Bladder {'gene': 607, 'blank': 3}\n",
      "Breast {'gene': 3215, 'blank': 38}\n",
      "Lymph node {'gene': 150, 'blank': 5}\n",
      "Liver {'gene': 2248, 'blank': 46}\n",
      "Heart {'gene': 378, 'blank': 37}\n",
      "Colon {'gene': 499, 'blank': 14}\n",
      "Prostate {'gene': 10556, 'blank': 55}\n",
      "Ovary {'gene': 2708, 'blank': 15}\n",
      "df_t len: 116801, df_join len:50808\n",
      "sample: cosmx_multitissue_htma\n",
      "23520085\n",
      "ct/spot/join:240154/23520085/7091176\n",
      "NSCLC {'neg_control_probe': 2888, 'gene': 1548531, 'sys_control': 80070}\n",
      "OvC {'neg_control_probe': 2372, 'gene': 1949535, 'sys_control': 30274}\n",
      "HNSCC {'neg_control_probe': 1381, 'gene': 685731, 'sys_control': 18923}\n",
      "BrC {'neg_control_probe': 1134, 'gene': 1577999, 'sys_control': 15227}\n",
      "Mel {'neg_control_probe': 745, 'gene': 475666, 'sys_control': 19121}\n",
      "CRC {'neg_control_probe': 700, 'gene': 672439, 'sys_control': 8440}\n",
      "df_t len: 23520085, df_join len:7091176\n",
      "sample: cosmx_multitissue_normal\n",
      "8878571\n",
      "ct/spot/join:178424/8878571/7002444\n",
      "Spleen {'neg_control_probe': 7222, 'gene': 1149617, 'sys_control': 183129}\n",
      "Prostate {'neg_control_probe': 565, 'gene': 388393, 'sys_control': 11877}\n",
      "Heart {'neg_control_probe': 575, 'gene': 116799, 'sys_control': 16587}\n",
      "Kidney {'neg_control_probe': 2369, 'gene': 377328, 'sys_control': 57565}\n",
      "Colon {'neg_control_probe': 1120, 'gene': 232438, 'sys_control': 19636}\n",
      "Skin {'neg_control_probe': 654, 'gene': 176527, 'sys_control': 11589}\n",
      "Tonsil {'neg_control_probe': 1772, 'gene': 2217213, 'sys_control': 29412}\n",
      "Liver {'neg_control_probe': 860, 'gene': 228191, 'sys_control': 12511}\n",
      "Thyroid {'neg_control_probe': 1093, 'gene': 987382, 'sys_control': 12877}\n",
      "Ovary {'neg_control_probe': 330, 'gene': 339298, 'sys_control': 3356}\n",
      "Lymph node {'neg_control_probe': 151, 'gene': 31477, 'sys_control': 5164}\n",
      "Breast {'neg_control_probe': 608, 'gene': 282044, 'sys_control': 6774}\n",
      "Bladder {'neg_control_probe': 212, 'gene': 63851, 'sys_control': 4430}\n",
      "Lung {'neg_control_probe': 72, 'gene': 17646, 'sys_control': 1730}\n",
      "df_t len: 8878571, df_join len:7002444\n",
      "sample: 2024_xenium_breast_htma\n",
      "sample: 2024_xenium_breast_tumor2\n",
      "sample: 2024_merscope_breast_htma\n",
      "data/2024_merscope_breast_htma/region_0/\n",
      "data/2024_merscope_breast_htma/region_1/\n",
      "data/2024_merscope_breast_htma/region_2/\n",
      "2569582\n",
      "NSCLC {'gene': 270805, 'blank': 2441}\n",
      "OvC {'gene': 948959, 'blank': 2994}\n",
      "Tonsil {'gene': 20074, 'blank': 440}\n",
      "Lymph node {'gene': 36230, 'blank': 201}\n",
      "Marker {'gene': 290, 'blank': 9}\n",
      "HNSCC {'gene': 74289, 'blank': 792}\n",
      "BrC {'gene': 323259, 'blank': 1624}\n",
      "Mel {'gene': 106305, 'blank': 1256}\n",
      "CRC {'gene': 130730, 'blank': 753}\n",
      "df_t len: 2048003, df_join len:1921451\n",
      "sample: 2024_merscope_breast_tumor2\n",
      "data/2024_merscope_breast_tumor2/region_0/\n",
      "data/2024_merscope_breast_tumor2/region_1/\n",
      "544237\n",
      "Mel {'gene': 6553, 'blank': 50}\n",
      "Marker normal liver {'gene': 6381, 'blank': 184}\n",
      "Kidney cancer {'gene': 758, 'blank': 3}\n",
      "Breast non-invasive DCIS_2 {'gene': 549, 'blank': 34}\n",
      "Lymphoma LN B cell {'gene': 45438, 'blank': 337}\n",
      "Marker normal spleen {'gene': 142, 'blank': 4}\n",
      "Breast invasive {'gene': 152, 'blank': 2}\n",
      "Testes cancer {'gene': 39737, 'blank': 129}\n",
      "SCC {'gene': 1978, 'blank': 16}\n",
      "Lymphoma non-Hodgkin {'gene': 1999, 'blank': 12}\n",
      "Lymphoma Hodgkin {'gene': 8990, 'blank': 41}\n",
      "Breast non-invasive DCIS_3 {'gene': 59}\n",
      "CRC {'gene': 14075, 'blank': 123}\n",
      "OvC {'gene': 10751, 'blank': 51}\n",
      "Liver cancer {'gene': 211625, 'blank': 583}\n",
      "Prostate cancer {'gene': 41820, 'blank': 259}\n",
      "BlC {'gene': 1856, 'blank': 86}\n",
      "Pancreas cancer {'gene': 6482, 'blank': 51}\n",
      "Renal cancer {'gene': 10041, 'blank': 101}\n",
      "Thyroid cancer {'gene': 27291, 'blank': 121}\n",
      "df_t len: 450664, df_join len:438864\n",
      "sample: 2024_cosmx_multitissue_htma\n",
      "99240555\n",
      "ct/spot/join:258986/99240555/26452348\n",
      "HNSCC {'neg_control_probe': 3791, 'gene': 2817393, 'sys_control': 39940}\n",
      "Mel {'neg_control_probe': 2143, 'gene': 1669830, 'sys_control': 22216}\n",
      "OvC {'neg_control_probe': 7683, 'gene': 8036720, 'sys_control': 77330}\n",
      "NSCLC {'neg_control_probe': 6741, 'gene': 5842049, 'sys_control': 120889}\n",
      "BrC {'neg_control_probe': 3177, 'gene': 4434857, 'sys_control': 41927}\n",
      "CRC {'neg_control_probe': 3006, 'gene': 3292301, 'sys_control': 30355}\n",
      "df_t len: 99240555, df_join len:26452348\n",
      "sample: 2024_cosmx_multitissue_tumor2\n",
      "81250185\n",
      "ct/spot/join:249682/81250185/68997583\n",
      "CRC {'neg_control_probe': 5289, 'gene': 4407995, 'sys_control': 69059}\n",
      "Breast invasive {'neg_control_probe': 1505, 'gene': 469166, 'sys_control': 26323}\n",
      "Breast non-invasive DCIS_2 {'neg_control_probe': 1428, 'gene': 878969, 'sys_control': 20013}\n",
      "Lymphoma non-Hodgkin {'neg_control_probe': 2203, 'gene': 2200846, 'sys_control': 39280}\n",
      "Renal cancer {'neg_control_probe': 3223, 'gene': 2052916, 'sys_control': 57674}\n",
      "Kidney cancer {'neg_control_probe': 2043, 'gene': 930613, 'sys_control': 42937}\n",
      "Breast non-invasive DCIS_3 {'neg_control_probe': 469, 'gene': 261638, 'sys_control': 3501}\n",
      "Pancreas cancer {'neg_control_probe': 3671, 'gene': 2937299, 'sys_control': 62390}\n",
      "Lymphoma Hodgkin {'neg_control_probe': 5986, 'gene': 6441191, 'sys_control': 59375}\n",
      "Lymphoma LN B cell {'neg_control_probe': 9403, 'gene': 11307223, 'sys_control': 145367}\n",
      "BlC {'neg_control_probe': 2724, 'gene': 2441193, 'sys_control': 55008}\n",
      "Prostate cancer {'neg_control_probe': 4033, 'gene': 4501536, 'sys_control': 50609}\n",
      "Testes cancer {'neg_control_probe': 4864, 'gene': 5545186, 'sys_control': 61359}\n",
      "Liver cancer {'neg_control_probe': 6254, 'gene': 7446140, 'sys_control': 66877}\n",
      "OvC {'neg_control_probe': 1982, 'gene': 1970688, 'sys_control': 31742}\n",
      "Thyroid cancer {'neg_control_probe': 6932, 'gene': 8483798, 'sys_control': 94493}\n",
      "SCC {'neg_control_probe': 1732, 'gene': 1703153, 'sys_control': 23594}\n",
      "Mel {'neg_control_probe': 3582, 'gene': 4011673, 'sys_control': 29436}\n",
      "df_t len: 81250185, df_join len:68997583\n"
     ]
    }
   ],
   "source": [
    "metric_by_sample = {}\n",
    "samples = []\n",
    "\n",
    "for SAMPLE in SAMPLES:\n",
    "\n",
    "    sample = SAMPLE.split('/')[1]\n",
    "    print (f'sample: {sample}')\n",
    "\n",
    "    if 'merscope' in sample:\n",
    "        neg_metric = 'blank'\n",
    "        df_metric = merscope_fdr(sample, neg_metric)\n",
    "    elif 'cosmx' in sample:\n",
    "        neg_metric = 'sys_control'\n",
    "        df_metric = cosmx_fdr(sample, neg_metric)\n",
    "\n",
    "    elif 'xenium' in sample:\n",
    "        neg_metric = 'blank'\n",
    "        df_metric = xenium_fdr(sample, neg_metric)\n",
    "    \n",
    "    metric_by_sample[sample] = df_metric\n",
    "    samples.append(sample)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['xenium_breast_htma', 'xenium_breast_normal', 'xenium_panhuman_htma', 'xenium_panhuman_normal', 'xenium_lung_htma', 'xenium_lung_normal', 'merscope_breast_htma', 'merscope_breast_normal', 'merscope_lung_htma', 'merscope_lung_normal', 'cosmx_multitissue_htma', 'cosmx_multitissue_normal', '2024_xenium_breast_htma', '2024_xenium_breast_tumor2', '2024_merscope_breast_htma', '2024_merscope_breast_tumor2', '2024_cosmx_multitissue_htma', '2024_cosmx_multitissue_tumor2']\n",
      "xenium_breast_htma\n",
      "xenium_breast_normal\n",
      "xenium_panhuman_htma\n",
      "xenium_panhuman_normal\n",
      "xenium_lung_htma\n",
      "xenium_lung_normal\n",
      "merscope_breast_htma\n",
      "merscope_breast_normal\n",
      "merscope_lung_htma\n",
      "merscope_lung_normal\n",
      "cosmx_multitissue_htma\n",
      "cosmx_multitissue_normal\n",
      "2024_xenium_breast_htma\n",
      "2024_xenium_breast_tumor2\n",
      "2024_merscope_breast_htma\n",
      "2024_merscope_breast_tumor2\n",
      "2024_cosmx_multitissue_htma\n",
      "2024_cosmx_multitissue_tumor2\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(f'{wd}/data/fdr'):\n",
    "    os.makedirs(f'{wd}/data/fdr')\n",
    "\n",
    "dfs = {}\n",
    "\n",
    "print (samples)\n",
    "\n",
    "for sample in samples:\n",
    "    df_metric = metric_by_sample[sample]\n",
    "    df_metric_no_marker = df_metric.loc[df_metric['tissue_type'].str.lower()!='marker']\n",
    "\n",
    "    if 'xenium' in sample:\n",
    "        df_metric_no_marker = df_metric_no_marker.groupby(by = ['tissue_type'], as_index=False).agg({\n",
    "            'fdr':'mean',\n",
    "            })\n",
    "\n",
    "    df_metric_no_marker['fdr'] = df_metric_no_marker['fdr'].apply(lambda x: round(x,2))\n",
    "    df_metric_no_marker['sample'] = sample\n",
    "    df_metric_no_marker['Platform_Panel'] = df_metric_no_marker['sample'].apply(lambda x: f\"{x.split('_')[-3].upper()}_{x.split('_')[-2].capitalize()}\")\n",
    "    df_metric_no_marker['modality'] = df_metric_no_marker['sample'].apply(lambda x: f\"{x.split('_')[-3].upper()}\")\n",
    "    df_metric_no_marker['panel'] = df_metric_no_marker['sample'].apply(lambda x: f\"{x.split('_')[-2].capitalize()}\")\n",
    "    dfs[sample] = df_metric_no_marker\n",
    "\n",
    "df_combo = pd.DataFrame()\n",
    "for sample in samples:\n",
    "    print (sample)\n",
    "    df_combo = pd.concat([df_combo,dfs[sample]])\n",
    "df_combo = correct_platform_panel(df_combo, 'Platform_Panel')\n",
    "df_combo.to_csv(f'{wd}/data/fdr/blank.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FDR Fig 2: Xenium (neg control probe) vs. CosMX (negative control)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample: xenium_breast_htma\n",
      "sample: xenium_breast_normal\n",
      "sample: xenium_panhuman_htma\n",
      "sample: xenium_panhuman_normal\n",
      "sample: xenium_lung_htma\n",
      "sample: xenium_lung_normal\n",
      "sample: merscope_breast_htma\n",
      "sample: merscope_breast_normal\n",
      "sample: merscope_lung_htma\n",
      "sample: merscope_lung_normal\n",
      "sample: cosmx_multitissue_htma\n",
      "23520085\n",
      "ct/spot/join:240154/23520085/7091176\n",
      "NSCLC {'neg_control_probe': 2888, 'gene': 1548531, 'sys_control': 80070}\n",
      "OvC {'neg_control_probe': 2372, 'gene': 1949535, 'sys_control': 30274}\n",
      "HNSCC {'neg_control_probe': 1381, 'gene': 685731, 'sys_control': 18923}\n",
      "BrC {'neg_control_probe': 1134, 'gene': 1577999, 'sys_control': 15227}\n",
      "Mel {'neg_control_probe': 745, 'gene': 475666, 'sys_control': 19121}\n",
      "CRC {'neg_control_probe': 700, 'gene': 672439, 'sys_control': 8440}\n",
      "df_t len: 23520085, df_join len:7091176\n",
      "sample: cosmx_multitissue_normal\n",
      "8878571\n",
      "ct/spot/join:178424/8878571/7002444\n",
      "Spleen {'neg_control_probe': 7222, 'gene': 1149617, 'sys_control': 183129}\n",
      "Prostate {'neg_control_probe': 565, 'gene': 388393, 'sys_control': 11877}\n",
      "Heart {'neg_control_probe': 575, 'gene': 116799, 'sys_control': 16587}\n",
      "Kidney {'neg_control_probe': 2369, 'gene': 377328, 'sys_control': 57565}\n",
      "Colon {'neg_control_probe': 1120, 'gene': 232438, 'sys_control': 19636}\n",
      "Skin {'neg_control_probe': 654, 'gene': 176527, 'sys_control': 11589}\n",
      "Tonsil {'neg_control_probe': 1772, 'gene': 2217213, 'sys_control': 29412}\n",
      "Liver {'neg_control_probe': 860, 'gene': 228191, 'sys_control': 12511}\n",
      "Thyroid {'neg_control_probe': 1093, 'gene': 987382, 'sys_control': 12877}\n",
      "Ovary {'neg_control_probe': 330, 'gene': 339298, 'sys_control': 3356}\n",
      "Lymph node {'neg_control_probe': 151, 'gene': 31477, 'sys_control': 5164}\n",
      "Breast {'neg_control_probe': 608, 'gene': 282044, 'sys_control': 6774}\n",
      "Bladder {'neg_control_probe': 212, 'gene': 63851, 'sys_control': 4430}\n",
      "Lung {'neg_control_probe': 72, 'gene': 17646, 'sys_control': 1730}\n",
      "df_t len: 8878571, df_join len:7002444\n",
      "sample: 2024_xenium_breast_htma\n",
      "sample: 2024_xenium_breast_tumor2\n",
      "sample: 2024_merscope_breast_htma\n",
      "sample: 2024_merscope_breast_tumor2\n",
      "sample: 2024_cosmx_multitissue_htma\n",
      "99240555\n",
      "ct/spot/join:258986/99240555/26452348\n",
      "HNSCC {'neg_control_probe': 3791, 'gene': 2817393, 'sys_control': 39940}\n",
      "Mel {'neg_control_probe': 2143, 'gene': 1669830, 'sys_control': 22216}\n",
      "OvC {'neg_control_probe': 7683, 'gene': 8036720, 'sys_control': 77330}\n",
      "NSCLC {'neg_control_probe': 6741, 'gene': 5842049, 'sys_control': 120889}\n",
      "BrC {'neg_control_probe': 3177, 'gene': 4434857, 'sys_control': 41927}\n",
      "CRC {'neg_control_probe': 3006, 'gene': 3292301, 'sys_control': 30355}\n",
      "df_t len: 99240555, df_join len:26452348\n",
      "sample: 2024_cosmx_multitissue_tumor2\n",
      "81250185\n",
      "ct/spot/join:249682/81250185/68997583\n",
      "CRC {'neg_control_probe': 5289, 'gene': 4407995, 'sys_control': 69059}\n",
      "Breast invasive {'neg_control_probe': 1505, 'gene': 469166, 'sys_control': 26323}\n",
      "Breast non-invasive DCIS_2 {'neg_control_probe': 1428, 'gene': 878969, 'sys_control': 20013}\n",
      "Lymphoma non-Hodgkin {'neg_control_probe': 2203, 'gene': 2200846, 'sys_control': 39280}\n",
      "Renal cancer {'neg_control_probe': 3223, 'gene': 2052916, 'sys_control': 57674}\n",
      "Kidney cancer {'neg_control_probe': 2043, 'gene': 930613, 'sys_control': 42937}\n",
      "Breast non-invasive DCIS_3 {'neg_control_probe': 469, 'gene': 261638, 'sys_control': 3501}\n",
      "Pancreas cancer {'neg_control_probe': 3671, 'gene': 2937299, 'sys_control': 62390}\n",
      "Lymphoma Hodgkin {'neg_control_probe': 5986, 'gene': 6441191, 'sys_control': 59375}\n",
      "Lymphoma LN B cell {'neg_control_probe': 9403, 'gene': 11307223, 'sys_control': 145367}\n",
      "BlC {'neg_control_probe': 2724, 'gene': 2441193, 'sys_control': 55008}\n",
      "Prostate cancer {'neg_control_probe': 4033, 'gene': 4501536, 'sys_control': 50609}\n",
      "Testes cancer {'neg_control_probe': 4864, 'gene': 5545186, 'sys_control': 61359}\n",
      "Liver cancer {'neg_control_probe': 6254, 'gene': 7446140, 'sys_control': 66877}\n",
      "OvC {'neg_control_probe': 1982, 'gene': 1970688, 'sys_control': 31742}\n",
      "Thyroid cancer {'neg_control_probe': 6932, 'gene': 8483798, 'sys_control': 94493}\n",
      "SCC {'neg_control_probe': 1732, 'gene': 1703153, 'sys_control': 23594}\n",
      "Mel {'neg_control_probe': 3582, 'gene': 4011673, 'sys_control': 29436}\n",
      "df_t len: 81250185, df_join len:68997583\n"
     ]
    }
   ],
   "source": [
    "metric_by_sample = {}\n",
    "samples = []\n",
    "\n",
    "for SAMPLE in SAMPLES:\n",
    "\n",
    "    sample = SAMPLE.split('/')[1]\n",
    "    print (f'sample: {sample}')\n",
    "\n",
    "    if 'xenium' in sample:\n",
    "        neg_metric = 'neg_control_probe'\n",
    "        df_metric = xenium_fdr(sample, neg_metric)\n",
    "\n",
    "        metric_by_sample[sample] = df_metric\n",
    "        samples.append(sample)\n",
    "    elif 'cosmx' in sample:\n",
    "        neg_metric = 'neg_control_probe'\n",
    "        df_metric = cosmx_fdr(sample, neg_metric)\n",
    "\n",
    "        metric_by_sample[sample] = df_metric\n",
    "        samples.append(sample)\n",
    "\n",
    "    # df_metric.core = df_metric.core + 200 if 'normal' in sample else df_metric.core"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['xenium_breast_htma', 'xenium_breast_normal', 'xenium_panhuman_htma', 'xenium_panhuman_normal', 'xenium_lung_htma', 'xenium_lung_normal', 'cosmx_multitissue_htma', 'cosmx_multitissue_normal', '2024_xenium_breast_htma', '2024_xenium_breast_tumor2', '2024_cosmx_multitissue_htma', '2024_cosmx_multitissue_tumor2']\n",
      "xenium_breast_htma\n",
      "xenium_breast_normal\n",
      "xenium_panhuman_htma\n",
      "xenium_panhuman_normal\n",
      "xenium_lung_htma\n",
      "xenium_lung_normal\n",
      "cosmx_multitissue_htma\n",
      "cosmx_multitissue_normal\n",
      "2024_xenium_breast_htma\n",
      "2024_xenium_breast_tumor2\n",
      "2024_cosmx_multitissue_htma\n",
      "2024_cosmx_multitissue_tumor2\n"
     ]
    }
   ],
   "source": [
    "dfs = {}\n",
    "\n",
    "print (samples)\n",
    "\n",
    "for sample in samples:\n",
    "    df_metric = metric_by_sample[sample]\n",
    "    df_metric_no_marker = df_metric.loc[df_metric['tissue_type'].str.lower()!='marker']\n",
    "\n",
    "    if 'xenium' in sample:\n",
    "        df_metric_no_marker = df_metric_no_marker.groupby(by = ['tissue_type'], as_index=False).agg({\n",
    "            'fdr':'mean',\n",
    "            })\n",
    "\n",
    "    df_metric_no_marker['fdr'] = df_metric_no_marker['fdr'].apply(lambda x: round(x,2))\n",
    "    df_metric_no_marker['sample'] = sample\n",
    "    df_metric_no_marker['Platform_Panel'] = df_metric_no_marker['sample'].apply(lambda x: f\"{x.split('_')[-3].upper()}_{x.split('_')[-2].capitalize()}\")\n",
    "    df_metric_no_marker['modality'] = df_metric_no_marker['sample'].apply(lambda x: f\"{x.split('_')[-3].upper()}\")\n",
    "    df_metric_no_marker['panel'] = df_metric_no_marker['sample'].apply(lambda x: f\"{x.split('_')[-2].capitalize()}\")\n",
    "    dfs[sample] = df_metric_no_marker\n",
    "\n",
    "df_combo = pd.DataFrame()\n",
    "for sample in samples:\n",
    "    print (sample)\n",
    "    df_combo = pd.concat([df_combo,dfs[sample]])\n",
    "    \n",
    "df_combo = correct_platform_panel(df_combo, 'Platform_Panel')\n",
    "df_combo.to_csv(f'{wd}/data/fdr/negative.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>fdr</th>\n",
       "      <th>sample</th>\n",
       "      <th>Platform_Panel</th>\n",
       "      <th>modality</th>\n",
       "      <th>panel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BrC</td>\n",
       "      <td>0.12</td>\n",
       "      <td>xenium_breast_htma</td>\n",
       "      <td>Xenium,breast</td>\n",
       "      <td>XENIUM</td>\n",
       "      <td>Breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CRC</td>\n",
       "      <td>0.20</td>\n",
       "      <td>xenium_breast_htma</td>\n",
       "      <td>Xenium,breast</td>\n",
       "      <td>XENIUM</td>\n",
       "      <td>Breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HNSCC</td>\n",
       "      <td>0.24</td>\n",
       "      <td>xenium_breast_htma</td>\n",
       "      <td>Xenium,breast</td>\n",
       "      <td>XENIUM</td>\n",
       "      <td>Breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mel</td>\n",
       "      <td>0.37</td>\n",
       "      <td>xenium_breast_htma</td>\n",
       "      <td>Xenium,breast</td>\n",
       "      <td>XENIUM</td>\n",
       "      <td>Breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NSCLC</td>\n",
       "      <td>0.16</td>\n",
       "      <td>xenium_breast_htma</td>\n",
       "      <td>Xenium,breast</td>\n",
       "      <td>XENIUM</td>\n",
       "      <td>Breast</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  tissue_type   fdr              sample Platform_Panel modality   panel\n",
       "0         BrC  0.12  xenium_breast_htma  Xenium,breast   XENIUM  Breast\n",
       "1         CRC  0.20  xenium_breast_htma  Xenium,breast   XENIUM  Breast\n",
       "2       HNSCC  0.24  xenium_breast_htma  Xenium,breast   XENIUM  Breast\n",
       "3         Mel  0.37  xenium_breast_htma  Xenium,breast   XENIUM  Breast\n",
       "4       NSCLC  0.16  xenium_breast_htma  Xenium,breast   XENIUM  Breast"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>fdr</th>\n",
       "      <th>sample</th>\n",
       "      <th>Platform_Panel</th>\n",
       "      <th>modality</th>\n",
       "      <th>panel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BrC</td>\n",
       "      <td>0.01</td>\n",
       "      <td>xenium_breast_htma</td>\n",
       "      <td>Xenium,breast</td>\n",
       "      <td>XENIUM</td>\n",
       "      <td>Breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CRC</td>\n",
       "      <td>0.02</td>\n",
       "      <td>xenium_breast_htma</td>\n",
       "      <td>Xenium,breast</td>\n",
       "      <td>XENIUM</td>\n",
       "      <td>Breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HNSCC</td>\n",
       "      <td>0.02</td>\n",
       "      <td>xenium_breast_htma</td>\n",
       "      <td>Xenium,breast</td>\n",
       "      <td>XENIUM</td>\n",
       "      <td>Breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mel</td>\n",
       "      <td>0.04</td>\n",
       "      <td>xenium_breast_htma</td>\n",
       "      <td>Xenium,breast</td>\n",
       "      <td>XENIUM</td>\n",
       "      <td>Breast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NSCLC</td>\n",
       "      <td>0.02</td>\n",
       "      <td>xenium_breast_htma</td>\n",
       "      <td>Xenium,breast</td>\n",
       "      <td>XENIUM</td>\n",
       "      <td>Breast</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  tissue_type   fdr              sample Platform_Panel modality   panel\n",
       "0         BrC  0.01  xenium_breast_htma  Xenium,breast   XENIUM  Breast\n",
       "1         CRC  0.02  xenium_breast_htma  Xenium,breast   XENIUM  Breast\n",
       "2       HNSCC  0.02  xenium_breast_htma  Xenium,breast   XENIUM  Breast\n",
       "3         Mel  0.04  xenium_breast_htma  Xenium,breast   XENIUM  Breast\n",
       "4       NSCLC  0.02  xenium_breast_htma  Xenium,breast   XENIUM  Breast"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combo = pd.read_csv(f'{wd}/data/fdr/blank.csv')\n",
    "df_combo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    200.000000\n",
       "mean       7.004900\n",
       "std       12.742925\n",
       "min        0.000000\n",
       "25%        0.030000\n",
       "50%        2.860000\n",
       "75%        6.850000\n",
       "max       71.540000\n",
       "Name: fdr, dtype: float64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combo['fdr'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['BrC', 'CRC', 'HNSCC', 'Mel', 'NSCLC', 'OvC', 'Bladder', 'Breast',\n",
       "       'Colon', 'Heart', 'Kidney', 'Liver', 'Lung', 'Lymph node', 'Ovary',\n",
       "       'Prostate', 'Skin', 'Spleen', 'Thyroid', 'Tonsil', 'BlC',\n",
       "       'Pancreas', 'Renal pelvis', 'Breast invasive',\n",
       "       'Breast non-invasive DCIS_2', 'Breast non-invasive DCIS_3',\n",
       "       'Kidney cancer', 'Liver cancer', 'Lymphoma Hodgkin',\n",
       "       'Lymphoma LN B cell', 'Lymphoma non-Hodgkin', 'Pancreas cancer',\n",
       "       'Prostate cancer', 'Renal cancer', 'SCC', 'Testes cancer',\n",
       "       'Thyroid cancer', 'Marker normal liver', 'Marker normal spleen'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_combo['tissue_type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
